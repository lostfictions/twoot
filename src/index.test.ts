import mockFs from "mock-fs";

import { twoot } from "./index";

let mastoMediaCreateCallCount = 0;
let mastoStatusCreateCallCount = 0;
jest.mock("masto", () => ({
  login: jest.fn(async ({ url }: { url: string }) => ({
    mediaAttachments: {
      create: jest.fn(async () => ({ id: `m${mastoMediaCreateCallCount++}` })),
    },
    statuses: {
      create: jest.fn(
        async ({
          status,
          visibility,
          inReplyToId,
          mediaIds,
        }: {
          status: string;
          visibility: string;
          inReplyToId: string;
          mediaIds?: string[];
        }) => {
          const id = `status_${mastoStatusCreateCallCount++}${
            inReplyToId ? `_inReplyTo[${inReplyToId}]` : ""
          }${mediaIds ? `_withMedia[${mediaIds.join(",")}]` : ""}`;

          return {
            id,
            visibility,
            content: status,
            uri: `${url}/${id}`,
            createdAt: "<time>",
          };
        }
      ),
    },
  })),
}));

let twitterMediaCreateCallCount = 0;
let twitterStatusCreateCallCount = 0;
jest.mock("twitter-api-client", () => ({
  TwitterClient: class {
    media = {
      mediaUpload: jest.fn(async () => ({
        media_id_string: `m${twitterMediaCreateCallCount++}`,
      })),
      mediaMetadataCreate: jest.fn(),
    };
    tweets = {
      statusesUpdate: jest.fn(
        async ({
          status,
          in_reply_to_status_id: replyId,
          media_ids,
        }: {
          status: string;
          in_reply_to_status_id?: string;
          media_ids?: string;
        }) => {
          const id = `status_${twitterStatusCreateCallCount++}${
            replyId ? `_inReplyTo[${replyId}]` : ""
          }${media_ids ? `_withMedia[${media_ids}]` : ""}`;

          return {
            id_str: id,
            text: status,
            created_at: "<time>",
            user: { screen_name: "username" },
          };
        }
      ),
    };
  },
}));

beforeEach(() => {
  mockFs();
});

afterEach(() => {
  mockFs.restore();
});

describe("api snapshots", () => {
  it("handles a simple status", async () => {
    const res = await twoot("hello world", [
      { type: "mastodon", server: "https://whatever.com", token: "xyz" },
      {
        type: "twitter",
        apiKey: "abc",
        apiSecret: "def",
        accessToken: "ghi",
        accessSecret: "jkl",
      },
    ]);

    mockFs.restore();
    expect(res).toMatchSnapshot();
  });

  it("handles a chain of statuses", async () => {
    const res = await twoot(
      [
        { status: "a twoot", media: [] },
        "a reply",
        { status: "a third one" },
        // mock-fs seems to be buggy, so we can't test media uploads as long as
        // masto... also buggily requires that we write to the filesystem :(
        // https://github.com/neet/masto.js/issues/481
        // https://github.com/tschaub/mock-fs/issues/338
        // { status: "another one", media: [{ buffer: Buffer.from("abcdef") }] },
      ],
      [
        {
          type: "twitter",
          apiKey: "abc",
          apiSecret: "def",
          accessToken: "ghi",
          accessSecret: "jkl",
        },
        { type: "mastodon", server: "https://whatever.com", token: "xyz" },
      ]
    );

    mockFs.restore();
    expect(res).toMatchSnapshot();
  });
});
