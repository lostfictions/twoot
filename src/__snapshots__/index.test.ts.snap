// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api snapshots handles a chain of statuses 1`] = `
[
  {
    "message": "a twoot
<time> => https://twitter.com/username/status/status_1
====
a reply
<time> => https://twitter.com/username/status/status_2_inReplyTo[status_1]
====
a third one
<time> => https://twitter.com/username/status/status_3_inReplyTo[status_2_inReplyTo[status_1]]",
    "statuses": [
      {
        "created_at": "<time>",
        "id_str": "status_1",
        "text": "a twoot",
        "user": {
          "screen_name": "username",
        },
      },
      {
        "created_at": "<time>",
        "id_str": "status_2_inReplyTo[status_1]",
        "text": "a reply",
        "user": {
          "screen_name": "username",
        },
      },
      {
        "created_at": "<time>",
        "id_str": "status_3_inReplyTo[status_2_inReplyTo[status_1]]",
        "text": "a third one",
        "user": {
          "screen_name": "username",
        },
      },
    ],
    "type": "twitter-chain",
  },
  {
    "message": "a twoot
<time> => https://whatever.com/status_1_withMedia[]
====
a reply
<time> => https://whatever.com/status_2_inReplyTo[status_1_withMedia[]]_withMedia[]
====
a third one
<time> => https://whatever.com/status_3_inReplyTo[status_2_inReplyTo[status_1_withMedia[]]_withMedia[]]_withMedia[]",
    "statuses": [
      {
        "content": "a twoot",
        "createdAt": "<time>",
        "id": "status_1_withMedia[]",
        "uri": "https://whatever.com/status_1_withMedia[]",
        "visibility": "public",
      },
      {
        "content": "a reply",
        "createdAt": "<time>",
        "id": "status_2_inReplyTo[status_1_withMedia[]]_withMedia[]",
        "uri": "https://whatever.com/status_2_inReplyTo[status_1_withMedia[]]_withMedia[]",
        "visibility": "public",
      },
      {
        "content": "a third one",
        "createdAt": "<time>",
        "id": "status_3_inReplyTo[status_2_inReplyTo[status_1_withMedia[]]_withMedia[]]_withMedia[]",
        "uri": "https://whatever.com/status_3_inReplyTo[status_2_inReplyTo[status_1_withMedia[]]_withMedia[]]_withMedia[]",
        "visibility": "public",
      },
    ],
    "type": "mastodon-chain",
  },
]
`;

exports[`api snapshots handles a simple status 1`] = `
[
  {
    "message": "hello world
<time> => https://whatever.com/status_0_withMedia[]",
    "status": {
      "content": "hello world",
      "createdAt": "<time>",
      "id": "status_0_withMedia[]",
      "uri": "https://whatever.com/status_0_withMedia[]",
      "visibility": "public",
    },
    "type": "mastodon",
  },
  {
    "message": "hello world
<time> => https://twitter.com/username/status/status_0",
    "status": {
      "created_at": "<time>",
      "id_str": "status_0",
      "text": "hello world",
      "user": {
        "screen_name": "username",
      },
    },
    "type": "twitter",
  },
]
`;
